packages/web-client/app/components/card-space/edit-details/url/index.ts

```diff
@@ -0,0 +1,84 @@
+import Component from '@glimmer/component';
+import { WorkflowCardComponentArgs } from '@cardstack/web-client/models/workflow';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { restartableTask, timeout } from 'ember-concurrency';
+import config from '@cardstack/web-client/config/environment';
+import { taskFor } from 'ember-concurrency-ts';
+import HubAuthentication from '@cardstack/web-client/services/hub-authentication';
+import { inject as service } from '@ember/service';
+import * as Sentry from '@sentry/browser';
+
+class CreateSpaceWorkflowEditDetailsUrlComponent extends Component<WorkflowCardComponentArgs> {
+  @service declare hubAuthentication: HubAuthentication;
+  @tracked subdomain: string = '';
+  @tracked url: string = '';
+  @tracked urlValidationState: 'valid' | 'invalid' | 'initial' | 'loading' =
+    'initial';
+  @tracked urlValidationMessage: string = '';
+
+  @action restoreFromSession() {
+    let url = this.args.workflowSession.getValue<string>('url');
+    if (url) {
+      this.subdomain = url.replace('.card.space', '');
+      this.url = url;
+      this.validateUrl();
+    }
+  }
+
+  @action onSubdomainInput(value: string) {
+    this.url = `${value}.card.space`;
+    this.validateUrl();
+  }
+
+  @action async validateUrl() {
+    try {
+      await taskFor(this.validateUrlTask).perform();
+    } catch (e) {
+      console.error(e);
+    }
+  }
+
+  @restartableTask *validateUrlTask(): any {
+    let url = this.url;
+    this.urlValidationState = 'loading';
+
+    yield timeout(config.environment === 'test' ? 10 : 500); // debounce
+
+    try {
+      let response = yield fetch(
+        `${config.hubURL}/api/card-spaces/validate-url`,
+        {
+          method: 'POST',
+          headers: {
+            Authorization: 'Bearer: ' + this.hubAuthentication.authToken,
+            Accept: 'application/vnd.api+json',
+            'Content-Type': 'application/vnd.api+json',
+          },
+          body: JSON.stringify({ data: { attributes: { url } } }),
+        }
+      );
+
+      let { errors } = yield response.json();
+
+      if (errors.length === 0) {
+        this.urlValidationState = 'valid';
+        this.urlValidationMessage = 'URL available';
+        this.args.workflowSession.setValue('url', url);
+      } else {
+        this.urlValidationState = 'invalid';
+        this.urlValidationMessage = errors[0].detail;
```

Comment:
Unfortunately [is-valid-domain](https://github.com/miguelmota/is-valid-domain) doesn't return error details.

Reply:
We could change the default message from 'Invalid URL' to 'Your Card Space subdomain can only contain letters, numbers, and hyphens'.

Reply:
And move the length constraint earlier, since length issues can return either `invalid URL` or `Max length is 50`. I guess these are not a must, since folks can check what a valid URL means by themselves, but providing good feedback seems nice.